### DISCUSSION TITLE
Difference between Provider and Signer.provider

### DISCUSSION THREAD

**Discussion by ackvf** [wevm/wagmi#1300] [Q&A]
Why is there a difference between `provider` and `signerProvider`?
```
const provider = useProvider()
const { data: signer } = useSigner()
const signerProvider = signer.provider
```

And what is the actual "official" difference?

---

One of the observable differences are that Provider makes network requests visible in Network tab in DevTools, while Signer.provider doesn't https://github.com/wagmi-dev/wagmi/discussions/1281

Another is, that In practice, using `provider` from `useProvider` causes various issues like 👇. Simply switching to `signer.provider` fixes it all.
> [Error: unknown account #0 (operation="getAddress", code=UNSUPPORTED_OPERATION)](https://stackoverflow.com/questions/71198438/ethers-js-error-unknown-account-0-operation-getaddress-code-unsupported-o)

**Comment by tmm** [comment]
> Why is there a difference between `provider` and `signerProvider`?

`provider` is used across wagmi for sending RPC requests. `signerProvider` isn't used for very much right now as it's not controlled by wagmi (at some point in the future, we may enable it as an option for sending RPC requests).

> And what is the actual "official" difference?

Often these are the same TypeScript type, but not the same instance/underlying implementation. `provider` from `useProvider` is [explicitly set by you](https://wagmi.sh/docs/client#provider) when configuring wagmi, while `provider` from `useSigner()?.provider` is set by the `Connector` (e.g. wallet app) with limited configuration ability.

> One of the observable differences are that Provider makes network requests visible in Network tab in DevTools, while Signer.provider doesn't https://github.com/wagmi-dev/wagmi/discussions/1281

`provider` makes requests as part of your code, while `signer.provider` uses the connected browser extension/wallet app.

> Another is, that In practice, using `provider` from `useProvider` causes various issues like

Happy to take a look if you create a [minimal reproducible example](
https://stackoverflow.com/help/minimal-reproducible-example)! (In general, it's better to use higher-level wagmi hooks (e.g. `useContractRead`) rather than manually making requests with `useProvider`. The higher-level hooks are declarative, handling request caching/deduplication, etc. Using `useProvider` is about as convenient as using raw ethers.)

**Reply by ackvf** [reply]
IIRC, the issue was happening when using the provider and requiring signing capabilities, i.e. requesting `provider.signer`. This would give an object where `{address: null, index: 0}` or smth like that. But this was fixed by using the actual signer from `useSigner` instead.

Anyway, we are actually unhappy that `provider` uses the definitions provided by wagmi. Even though we can set our own definitions, it's still not optimal as the user might prefer to use their own (e.g. paid Alchemy) faster RPC endpoints. That means we prefer whatever is injected by the connector.

At one point, we were using the `allChains` provided by wagmi. We would prompt the user to add new chain from our definitions, but then when that user switched network, it would not use our definitions, which puzzled us a lot, until we realized how it actually works under the hood.
